File: src/main/java/aurocosh/divinefavor/common/entity/minions/base/MinionData.java
package aurocosh.divinefavor.common.entity.minions.base;

import com.google.common.base.Optional;
import net.minecraft.entity.EntityCreature;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.network.datasync.DataParameter;
import net.minecraft.network.datasync.EntityDataManager;

import javax.annotation.Nullable;
import java.util.UUID;

public class MinionData<T extends EntityCreature & IMinion> {
    private static final String TAG_OWNER_UUID = "OwnerUUID";
    private static final String TAG_MODE = "Mode";

    private final T minion;
    private final EntityDataManager dataManager;

    private final DataParameter<Boolean> begging;
    private final DataParameter<Integer> mode;
    private final DataParameter<Optional<UUID>> ownerUniqueId;

    private EntityLivingBase attackTarget;

    public MinionData(T minion, EntityDataManager dataManager, DataParameter<Boolean> begging, DataParameter<Integer> mode, DataParameter<Optional<UUID>> ownerUniqueId) {
        this.minion = minion;
        this.dataManager = dataManager;
        this.begging = begging;
        this.mode = mode;
        this.ownerUniqueId = ownerUniqueId;
    }

    public MinionMode getMode() {
        return MinionMode.VALUES[dataManager.get(mode)];
    }

    public void setMode(MinionMode minionMode) {
        dataManager.set(mode, minionMode.getValue());
    }

    public boolean isBegging() {
        return dataManager.get(begging);
    }

    public void setBegging(boolean beg) {
        dataManager.set(begging, beg);
    }

    @Nullable
    public UUID getOwnerUUID() {
        return (UUID) ((Optional) dataManager.get(ownerUniqueId)).orNull();
    }

    public void setOwnerUUID(UUID ownerUUID) {
        dataManager.set(ownerUniqueId, Optional.fromNullable(ownerUUID));
    }

    @Nullable
    public EntityPlayer getOwner() {
        try {
            UUID uuid = getOwnerUUID();
            return uuid == null ? null : minion.world.getPlayerEntityByUUID(uuid);
        } catch (IllegalArgumentException var2) {
            return null;
        }
    }

    @Nullable
    public void setOwner(EntityPlayer owner) {
        setOwnerUUID(owner.getGameProfile().getId());
    }

    @Nullable
    public boolean hasOwner() {
        return (dataManager.get(ownerUniqueId)).isPresent();
    }

    public boolean isOwner(EntityLivingBase entityIn) {
        return entityIn == getOwner();
    }

    public void setOwnerByUUID(String ownerUUID) {
        setOwnerUUID(ownerUUID.isEmpty() ? null : UUID.fromString(ownerUUID));
    }

    public void writeEntityToNBT(NBTTagCompound compound) {
        UUID ownerId = getOwnerUUID();
        compound.setString(TAG_OWNER_UUID, ownerId == null ? "" : ownerId.toString());
        compound.setInteger(TAG_MODE, getMode().getValue());
    }

    public void readEntityFromNBT(NBTTagCompound compound) {
        setOwnerByUUID(compound.getString(TAG_OWNER_UUID));
        setMode(MinionMode.fromInt(compound.getInteger(TAG_MODE)));
    }

    public EntityLivingBase getAttackTarget() {
        return attackTarget;
    }

    public void setAttackTarget(EntityLivingBase livingBase){
        attackTarget = livingBase;
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/entity/minions/base/MinionData.kt
package aurocosh.divinefavor.common.entity.minions.base

import com.google.common.base.Optional
import net.minecraft.entity.EntityCreature
import net.minecraft.entity.EntityLiving
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.network.datasync.DataParameter
import net.minecraft.network.datasync.EntityDataManager
import java.util.*

class MinionData<T>(private val minion: T, private val dataManager: EntityDataManager, private val beggingParameter: DataParameter<Boolean>, private val modeParameter: DataParameter<Int>, private val ownerUniqueIdParameter: DataParameter<Optional<UUID>>) where T : IMinion, T : EntityCreature {

    var attackTarget: EntityLivingBase? = null

    var isBegging: Boolean
        get() = dataManager.get(beggingParameter)
        set(beg) = dataManager.set(beggingParameter, beg)

    var ownerUUID: UUID?
        get() = (dataManager.get(ownerUniqueIdParameter) as Optional<*>).orNull() as UUID?
        set(ownerUUID) = dataManager.set(ownerUniqueIdParameter, Optional.fromNullable(ownerUUID))

    var owner: EntityPlayer?
        get() {
            try {
                val uuid = ownerUUID
                return if (uuid == null) null else minion.world.getPlayerEntityByUUID(uuid)
            } catch (var2: IllegalArgumentException) {
                return null
            }
        }
        set(owner) {
            ownerUUID = owner?.gameProfile?.id
        }

    val isFollowing: Boolean
        get() = mode === MinionMode.Follow

    val isWaiting: Boolean
        get() = mode === MinionMode.Wait

    var mode: MinionMode = MinionMode.Normal
        get() {
            return MinionMode.INDEXER[dataManager.get(modeParameter)]
        }
        set(value) {
            field = value
            dataManager.set(modeParameter, value.value)
        }

    fun hasOwner(): Boolean {
        return dataManager.get(ownerUniqueIdParameter).isPresent
    }

    fun isOwner(entityIn: EntityLivingBase): Boolean {
        return entityIn === owner
    }

    fun setOwnerByUUID(ownerUUID: String) {
        this.ownerUUID = if (ownerUUID.isEmpty()) null else UUID.fromString(ownerUUID)
    }

    fun writeEntityToNBT(compound: NBTTagCompound) {
        val ownerId = ownerUUID
        compound.setString(TAG_OWNER_UUID, ownerId?.toString() ?: "")
        compound.setInteger(TAG_MODE, mode.value)
    }

    fun readEntityFromNBT(compound: NBTTagCompound) {
        setOwnerByUUID(compound.getString(TAG_OWNER_UUID))
        mode = MinionMode.INDEXER[compound.getInteger(TAG_MODE)]
    }

    fun shouldAttack(entity: EntityLiving?): Boolean {
        val owner = owner
        if (entity === owner)
            return false
        if (entity is IMinion) {
            val minion = entity as IMinion
            if (minion.minionData.owner === owner)
                return false
        }
        return (entity as? EntityLivingBase)?.attackable() ?: false
    }

    companion object {
        private val TAG_OWNER_UUID = "OwnerUUID"
        private val TAG_MODE = "Mode"
    }
}
